timemachine/kernels/wrap_kernels.cpp

```diff
@@ -0,0 +1,101 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/numpy.h>
+
+#include "gpu/potential.hpp"
+#include "gpu/custom_bonded_gpu.hpp"
+
+namespace py = pybind11;
+
+
+template <typename RealType>
+void declare_potential(py::module &m, const char *typestr) {
+
+    using Class = timemachine::Potential<RealType>;
+    std::string pyclass_name = std::string("Potential") + typestr;
+    py::class_<Class>(m, pyclass_name.c_str(), py::buffer_protocol(), py::dynamic_attr());
+
+}
+
+#include<iostream>
+
+template<typename RealType>
+void declare_harmonic_bond(py::module &m, const char *typestr) {
+
+    using Class = timemachine::HarmonicBond<RealType>;
+    std::string pyclass_name = std::string("HarmonicBond_") + typestr;
+    py::class_<Class, timemachine::Potential<RealType> >(
+        m,
+        pyclass_name.c_str(),
+        py::buffer_protocol(),
+        py::dynamic_attr()
+    )
+    .def(py::init([](
+        const py::array_t<int, py::array::c_style> &bi, // bond_idxs
+        const py::array_t<int, py::array::c_style> &pi// param_idxs
+    ) {
+        std::vector<int> bond_idxs(bi.size());
+        std::memcpy(bond_idxs.data(), bi.data(), bi.size()*sizeof(int));
+        std::vector<int> param_idxs(pi.size());
+        std::memcpy(param_idxs.data(), pi.data(), pi.size()*sizeof(int));
+        return new timemachine::HarmonicBond<RealType>(bond_idxs, param_idxs);
+    }))
+    .def("derivatives", [](timemachine::HarmonicBond<RealType> &nrg,
+        const py::array_t<RealType, py::array::c_style> &coords,
+        const py::array_t<RealType, py::array::c_style> &params,
+        const py::array_t<RealType, py::array::c_style> &dxdps
+        ) -> py::tuple {
+
+        const long unsigned int num_atoms = coords.shape()[0];
+        const long unsigned int num_dims = coords.shape()[1];
+        const long unsigned int num_params = params.shape()[0];
+
+        py::array_t<RealType, py::array::c_style> py_E({1});
+        py::array_t<RealType, py::array::c_style> py_dE_dp({num_params});
+        py::array_t<RealType, py::array::c_style> py_dE_dx({num_atoms, num_dims});
+        py::array_t<RealType, py::array::c_style> py_d2E_dxdp({num_params, num_atoms, num_dims});
+
+        memset(py_E.mutable_data(), 0.0, sizeof(RealType));
+        memset(py_dE_dp.mutable_data(), 0.0, sizeof(RealType)*num_params);
+        memset(py_dE_dx.mutable_data(), 0.0, sizeof(RealType)*num_atoms*num_dims);
+        memset(py_d2E_dxdp.mutable_data(), 0.0, sizeof(RealType)*num_params*num_atoms*num_dims);
+
+        RealType *dxdps_ptr = nullptr;
+
+        if(dxdps.size() > 0) {
+
+            // this is dangerously safe since it then gets passed into a function
+            // that takes in a const RealType * again
+            dxdps_ptr = const_cast<RealType *>(dxdps.data());
+            std::cout << "SETTING ACTUAL DATA" << dxdps.size() << " " << dxdps.shape() << " " << dxdps.data()[0] << std::endl;
```

Comment:
@bp-kelley 
what's weird is that if I pass in a None into a np array it actually turns into a size 1 scalar of value nan :)

Reply:
I think I need to denote None in numpy land by doing something like:

np.zeros(shape=(0,)) # this has size 0